src/main/scala/ru/ledostuff/calories/infrastructure/services/ProductCaloriesService.scala

```diff
@@ -42,23 +42,23 @@ class ProductCaloriesService[F[_]: Sync](productService: ProductService[F],
     ).getOrElseF({
       val resultOrError: EitherT[F, ErrorResponse, ProductCaloriesInfo] = for {
 
+        _                     <- EitherT.liftF(log.info(s"search translation for name $name"))
         translatedProductInfo <- EitherT.fromOptionF[F, ErrorResponse, TranslatedProduct]({
-          (OptionT.liftF(log.info(s"search translation for name $name")) *>
-          translateService.translateProductName(name)).value
+          translateService.translateProductName(name).value
         }, TranslationNotFound)
 
-        savedProduct          <- EitherT.liftF[F, ErrorResponse, Product](productService.createProduct(Product(name, translatedProductInfo.translatedNames.toSeq)))
+        savedProduct          <- EitherT.liftF[F, ErrorResponse, Product](productService.createProduct(Product(name, translatedProductInfo.translatedNames.toList)))
 
         caloriesInfo          <- {
-          val engNamesRequests = savedProduct.engNames.toList.map { engName =>
+          val engNamesRequests = savedProduct.engNames.map { engName =>
             (EitherT.liftF[F, ErrorResponse, Unit](log.info(s"search calories for name $engName")) *>
               EitherT.fromOptionF[F, ErrorResponse, ProductCalories](productCaloriesService.getCaloriesByProductName(engName).value, ProductCaloriesNotFound)).value
           }
           EitherT.liftF(engNamesRequests.sequence.map(l => l.collect{case Right(e) => e})).ensure(ProductCaloriesNotFound)(_.nonEmpty)
```

Comment:
Why didn't you use traverse instead of sequence?

Reply:
`
EitherT.liftF({engNamesRequests.traverse(e => e).map(_.collect({case Right(e) => e}))}).ensure(ProductCaloriesNotFound)(_.nonEmpty)
`
For me it looks the same. Could we make a filter in one go?

Reply:
I meant this:
```scala
val engNamesRequests = savedProduct.engNames.traverse { engName =>
  (EitherT.liftF[F, ErrorResponse, Unit](log.info(s"search calories for name $engName")) *>
    EitherT.fromOptionF[F, ErrorResponse, ProductCalories](productCaloriesService.getCaloriesByProductName(engName).value, ProductCaloriesNotFound)).value
}
EitherT.liftF(engNamesRequests.map(l => l.collect{case Right(e) => e})).ensure(ProductCaloriesNotFound)(_.nonEmpty)
```

We can simplify it:
```scala
val engNamesRequests = savedProduct.engNames.traverse { engName =>
  log.info(s"search calories for name $engName") *>
  productCaloriesService.getCaloriesByProductName(engName).value
}
EitherT.liftF(engNamesRequests.map(l => l.collect{case Some(e) => e})).ensure(ProductCaloriesNotFound)(_.nonEmpty)
```

And finally, you are able to use a special version of `traverse` - `traverseFilter`:
```scala
val engNamesRequests = savedProduct.engNames.traverseFilter { engName =>
  (log.info(s"search calories for name $engName") *>
  productCaloriesService.getCaloriesByProductName(engName).value)
}
EitherT.liftF(engNamesRequests).ensure(ProductCaloriesNotFound)(_.nonEmpty)
```

Reply:
Moreover, you have an opportunity to encode "non-emptiness" on a type-level.

```scala
val engNamesRequests = savedProduct.engNames.traverseFilter { engName =>
log.info(s"search calories for name $engName") *>
  productCaloriesService.getCaloriesByProductName(engName).value
}.map(NonEmptyList.fromList)
EitherT.fromOptionF(engNamesRequests, ProductCaloriesNotFound)
```